src/main/resources/frontend/src/views/AdminView/SubmissionInfo.vue

```diff
@@ -50,14 +55,18 @@ const rowData = reactive({
         <span v-if="submission.passed">Submission passed!</span>
         <span v-else class="failure">Submission failed</span>
       </p>
+      <div v-if="!submission.approved" class="blocked-submission-notice">
+        <p>This submission has been blocked!</p>
+        <button>Approve with penalty</button>
```

Comment:
@frozenfrank  I can write the actual endpoint for sure. 

It looks like this is expecting the front end to do the calculation for the score. Is that what we want? How much control is the TA supposed to have? My understanding is that the TA has two options: Approve with a penalty, or approve and waive the penalty. 

Should approveSubmission just take in a boolean `withPenalty` and it holds the logic for what that means? Or should the front end have the logic that tells the server what percent to take off?

Reply:
@DallinFromEarth This `approveSubmission()` method is designed to be flexible and support either approach. The key is to notice the parameter `@Nullable Float approvedScore` is a **nullable** field. 

If you leave that blank, then the server will do the math based on a fixed algorithm. Currently, that algorithm is defined as:
1. Identify the most recent passing score for the student on the given phase
2. Reduce the score as follows: `score = mostRecentScore * (1 - penaltyPct)`

Then the rest of the details and the way you send that information between the client and server before you call this function are totally up to you. Here are several options:
1. The TA clicks "approve with penalty" on a specific submission. You deduct the points from that submission and send a pre-calculated `approvedScore` field along with the `penaltyPct` field set to the appropriate value (likely 10%). This translates directly into an endpoint and the provided server function. This approach is only as secure as the security of admin logins. Someone with an admin login could set the field to any value they want.
2. The TA clicks "approve with penalty." The client hits a ".../with-penalty" endpoint with fields `studentNetId` and `phase` and `approverNetId`. The endpoint then looks up the penalty from a config table (secure) and passes in the function like this:
    - `studentNetId = studentNetId`
    - `phase = phase`
    - `approverNetId = approverNetId`
    - `approvedScore = null`
    - `penaltyPct = officialPenaltyPercentageFromDao`
3. The TA clicks "approve without penalty." The client hits a `".../without-penalty" endpoint with all the same fields as the previous case. This time, it's not necessary to look up any values. The `approveSubmission()` method can be called with the following parameters:
    - `studentNetId = studentNetId`
    - `phase = phase`
    - `approverNetId = approverNetId`
    - `approvedScore = null`
    - `penaltyPct = 0`
4. The TA choose a whether or not to apply a penalty, and sends the minimal required information to the server. The server then uses a _a different heuristic_ (the `approveSubmission()` method only chooses the most recent one, but we may want to use the first submission, or something else etc) to choose a score which will be the basis for the penalty. The server does the math and passes in the final approved score to the `approveSubmission()` method.
5. A professor wants to approve the submission, but has additional information and wants to give a different score. He may reduce more than 10% because it's a repeat offense, or he may only reduce 5% because he's feeling generous. Maybe he detects cheating and wants to set the score to a specific value (0). This value can be passed to the server and made effective. Or, the professor may choose to base the score not on the most recent submission (maybe it came after the deadline), but only the last one before the deadline.

The point is that this architecture leaves the door open to lots of options. You are building the bridge between the client and the server, and while there are some minimum requirements, you can design the rest of it in any of multiple valid ways.